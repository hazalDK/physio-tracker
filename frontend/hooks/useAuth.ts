import * as SecureStore from "expo-secure-store";
import axios from "axios";
import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamsList } from "@/types/navigation";
import { Alert } from "react-native";

export function useAuth() {
  const navigation =
    useNavigation<StackNavigationProp<RootStackParamsList, "login">>();

  const getToken = async () => {
    return await SecureStore.getItemAsync("access_token");
  };

  const refreshToken = async () => {
    try {
      const refreshToken = await SecureStore.getItemAsync("refresh_token");
      if (!refreshToken) {
        throw new Error("No refresh token available");
      }

      const refreshUrl = process.env.API_URL || "http://192.168.68.111:8000";
      const refreshResponse = await axios.post(
        `${refreshUrl}/api/token/refresh/`,
        { refresh: refreshToken }
      );

      const newToken = refreshResponse.data.access;
      const newRefreshToken = refreshResponse.data.refresh;

      await Promise.all([
        SecureStore.setItemAsync("access_token", newToken),
        SecureStore.setItemAsync("refresh_token", newRefreshToken),
      ]);

      return newToken;
    } catch (error) {
      await Promise.all([
        SecureStore.deleteItemAsync("access_token"),
        SecureStore.deleteItemAsync("refresh_token"),
      ]);

      Alert.alert("Session Expired", "Please login again", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);

      return null;
    }
  };

  const createApiInstance = async () => {
    let token = await getToken();

    if (!token) {
      Alert.alert("Login Required", "Please sign in to continue", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);
      return null;
    }

    return axios.create({
      baseURL: process.env.API_URL || "http://192.168.68.111:8000",
      timeout: 10000,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
  };

  return { getToken, refreshToken, createApiInstance };
}
