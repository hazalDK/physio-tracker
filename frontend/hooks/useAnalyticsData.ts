import { useState, useEffect } from "react";
import axios from "axios";
import { useAuth } from "./useAuth";

// Helper function to format date
const formatDate = (date: Date) => {
  return date.toISOString().split("T")[0];
};

// Type for analytics data (can be either adherence or pain data)
interface ChartData {
  labels: string[];
  datasets: {
    data: number[];
  }[];
}

/**
 * Custom hook to fetch and manage analytics data (adherence or pain levels).
 * @param {string} endpoint - The API endpoint to fetch data from.
 * @param {"adherence" | "pain"} dataType - The type of data to fetch (either adherence or pain).
 * @returns {Object} - An object containing loading state, chart data, average, history, error, and functions to navigate weeks.
 * @throws {Error} - Throws an error if the API call fails.
 * */
export function useAnalyticsData<T>(
  endpoint: string,
  dataType: "adherence" | "pain"
): object {
  const [loading, setLoading] = useState(true);
  const [chartData, setChartData] = useState<ChartData>({
    labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
    datasets: [{ data: [0, 0, 0, 0, 0, 0, 0] }],
  });
  const [average, setAverage] = useState(0);
  const [history, setHistory] = useState<T[]>([]);
  const [error, setError] = useState<string | null>(null);

  // Add week navigation state
  const [currentEndDate, setCurrentEndDate] = useState(new Date());
  const [weekTitle, setWeekTitle] = useState("Current Week");
  const { createApiInstance, refreshToken } = useAuth();

  // Function to go to previous week
  const goToPreviousWeek = () => {
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() - 7);
    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Function to go to next week (but not beyond current date)
  const goToNextWeek = () => {
    const today = new Date();
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() + 7);

    // Don't go beyond current date
    if (newEndDate > today) {
      newEndDate.setTime(today.getTime());
    }

    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Update week title based on date range
  const updateWeekTitle = (endDate: Date) => {
    const startDate = new Date(endDate);
    startDate.setDate(startDate.getDate() - 6);

    const today = new Date();
    const isCurrentWeek =
      today.getDate() === endDate.getDate() &&
      today.getMonth() === endDate.getMonth() &&
      today.getFullYear() === endDate.getFullYear();

    if (isCurrentWeek) {
      setWeekTitle("Current Week");
    } else {
      const startMonth = startDate.toLocaleString("default", {
        month: "short",
      });
      const endMonth = endDate.toLocaleString("default", { month: "short" });

      if (startMonth === endMonth) {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endDate.getDate()}`
        );
      } else {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endMonth} ${endDate.getDate()}`
        );
      }
    }
  };

  // Function to fetch data for a specific end date
  const fetchData = async (endDate: Date) => {
    setLoading(true);
    try {
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch stats from the API with date parameter
        const response = await api.get(
          `${endpoint}?end_date=${formatDate(endDate)}`
        );
        if (response.data) {
          setChartData(response.data.chart_data);

          if (dataType === "adherence") {
            setAverage(response.data.average_adherence);
          } else {
            setAverage(response.data.average_pain);
          }

          setHistory(response.data.history);
        }
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) return;

          const api = await createApiInstance();
          if (!api) return;

          // Retry with new token
          const response = await api.get(
            `${endpoint}?end_date=${formatDate(endDate)}`
          );

          if (response.data) {
            setChartData(response.data.chart_data);

            if (dataType === "adherence") {
              setAverage(response.data.average_adherence);
            } else {
              setAverage(response.data.average_pain);
            }

            setHistory(response.data.history);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          setError(
            `Failed to load your ${
              dataType === "adherence" ? "exercises" : "pain data"
            }. Please check your connection and try again.`
          );
        }
      }
    } catch (error) {
      console.error(
        `Failed to load your ${
          dataType === "adherence" ? "exercises" : "pain data"
        }.`
      );
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    // Initial data fetch for current week
    fetchData(currentEndDate);
    updateWeekTitle(currentEndDate);
  }, []);

  return {
    loading,
    chartData,
    average,
    history,
    error,
    weekTitle,
    goToPreviousWeek,
    goToNextWeek,
  };
}
