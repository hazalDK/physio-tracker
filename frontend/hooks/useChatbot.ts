import { useState, useEffect, useCallback } from "react";
import { Alert } from "react-native";
import Toast from "react-native-root-toast";
import { messageType } from "@/types/chatbot";
import { createChatApi, apiRequestWithRefresh } from "../api";
import { useAuth } from "./useAuth";

export default function useChatbot() {
  const auth = useAuth();
  const [messages, setMessages] = useState<messageType[]>([]);
  const [userExercises, setUserExercises] = useState([]);
  const [recentPain, setRecentPain] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [isInitialLoading, setIsInitialLoading] = useState(true);
  const [api, setApi] = useState<any>(null);

  // Initialize API and chat
  useEffect(() => {
    const initializeApi = async () => {
      const apiInstance = await auth.createApiInstance();
      setApi(apiInstance);
    };

    initializeApi();

    // Set initial messages
    setMessages([
      {
        id: "disclaimer",
        text: "DISCLAIMER",
        sender: "disclaimer",
      },
      {
        id: "welcome",
        text: "Welcome to your physiotherapy assistant! How can I help you today?",
        sender: "bot",
      },
    ]);
  }, []);

  // Fetch user data when API is available
  useEffect(() => {
    if (api) {
      fetchUserData();
    }
  }, [api]);

  const fetchUserData = async () => {
    if (!api) return;

    setIsInitialLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const fetchData = async () => {
        const [reportsData, exercisesData] = await Promise.all([
          chatApi.fetchReports(),
          chatApi.fetchUserExercises(),
        ]);

        return { reportsData, exercisesData };
      };

      const { reportsData, exercisesData } = await apiRequestWithRefresh(
        fetchData,
        auth
      );

      setUserExercises(exercisesData);

      if (reportsData.length > 0) {
        setRecentPain(reportsData[0].pain_level);
      }
    } catch (error) {
      console.error("Failed to fetch user data:", error);

      Alert.alert(
        "Error",
        "Failed to load your exercises. Please check your connection and try again."
      );
    } finally {
      setIsInitialLoading(false);
    }
  };

  const sendMessage = async (inputText: string) => {
    if (!inputText.trim() || !api) return;

    // Add user message to chat
    const userMessageId = Date.now().toString();
    setMessages((prev) => [
      ...prev,
      { id: userMessageId, text: inputText, sender: "user" },
    ]);

    setIsLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Create context object with user's exercise data
      const exerciseContext = {
        exercises: userExercises,
        recentPain: recentPain,
      };

      // Use the wrapper function for API calls with refresh
      const sendMessageCall = async () => {
        return await chatApi.sendChatMessage(inputText, exerciseContext);
      };

      // Send message to API with refresh handling
      const response = await apiRequestWithRefresh(sendMessageCall, auth);

      // Add bot response to chat
      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: response.message,
          sender: "bot",
        },
      ]);
    } catch (error) {
      console.error("Error sending message:", error);

      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: "Sorry, I encountered an error. Please try again.",
          sender: "bot",
        },
      ]);
    } finally {
      setIsLoading(false);
    }
  };

  const resetChat = async () => {
    if (!api) return;

    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const resetCall = async () => {
        await chatApi.resetChat();
      };

      await apiRequestWithRefresh(resetCall, auth);

      // Reset messages
      setMessages([
        {
          id: "disclaimer",
          text: "DISCLAIMER",
          sender: "disclaimer",
        },
        {
          id: "welcome",
          text: "Welcome to your physiotherapy assistant! How can I help you today?",
          sender: "bot",
        },
      ]);

      // Refresh user data
      fetchUserData();

      Toast.show("Chat reset!", {
        duration: Toast.durations.SHORT,
        position: Toast.positions.BOTTOM,
        backgroundColor: "#14b8a6",
        textColor: "white",
        shadow: true,
      });
    } catch (error) {
      console.error("Failed to reset chat:", error);
      Alert.alert("Error", "Could not reset chat.");
    }
  };

  return {
    messages,
    isLoading,
    isInitialLoading,
    sendMessage,
    resetChat,
  };
}
