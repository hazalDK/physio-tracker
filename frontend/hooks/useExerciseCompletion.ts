import { useState } from "react";
import { Alert } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { useAuth } from "./useAuth";

// Custom hook to manage user profile data and related operations.
export function useExerciseCompletion(userExerciseId: number | null) {
  const [showCompletionForm, setShowCompletionForm] = useState(false);
  const [reps, setReps] = useState(0);
  const [sets, setSets] = useState(0);
  const [painLevel, setPainLevel] = useState(0);
  const navigation = useNavigation();
  const { createApiInstance } = useAuth();

  // Function to handle saving exercise completion data
  // This function is called when the user submits the completion form.
  const handleSave = async () => {
    // Validate input fields
    if (!reps || !sets || !painLevel) {
      Alert.alert("Error", "Please fill all required fields");
      return;
    }

    // Api instance creation
    const api = await createApiInstance();
    if (!api) return;

    try {
      // Update user exercise data
      await api.put(`/user-exercises/${userExerciseId}/`, {
        reps: reps,
        sets: sets,
        pain_level: painLevel,
        completed: true,
      });

      // Closes the modal
      setShowCompletionForm(false);

      // Handle pain level logic
      // If pain level is low (1-3), check if they can increase difficulty
      if (painLevel <= 3) {
        try {
          const canIncreaseResponse = await api.get(
            `/user-exercises/${userExerciseId}/can_increase/`
          );

          if (canIncreaseResponse.data.can_increase) {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low and you've had consistent low pain for 3 days. Would you like to increase the difficulty for next time?",
              [
                {
                  text: "No",
                  onPress: () => navigation.goBack(),
                  style: "cancel",
                },
                {
                  text: "Yes",
                  onPress: async () => {
                    try {
                      await api.post(
                        `/user-exercises/${userExerciseId}/confirm_increase/`,
                        { confirm: "yes" }
                      );
                      Alert.alert(
                        "Success",
                        "Exercise difficulty increased for next time!"
                      );
                      navigation.goBack();
                    } catch (error) {
                      console.error("Error increasing difficulty:", error);
                      Alert.alert(
                        "Error",
                        "Failed to increase difficulty. Please try again."
                      );
                      navigation.goBack();
                    }
                  },
                },
              ]
            );
          } else {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low. Keep it up for 3 days to unlock higher difficulty!",
              [{ text: "OK", onPress: () => navigation.goBack() }]
            );
          }
        } catch (error) {
          // Handle error when checking increase eligibility
          console.error("Error checking increase eligibility:", error);
          Alert.alert("Success", "Exercise completed successfully", [
            { text: "OK", onPress: () => navigation.goBack() },
          ]);
        }
      } else if (painLevel > 3) {
        // If pain level is high, ask if they want to decrease difficulty
        Alert.alert(
          "Exercise Difficulty",
          "Your pain level is high. Would you like to decrease the difficulty for next time?",
          [
            {
              text: "No",
              onPress: () => navigation.goBack(),
              style: "cancel",
            },
            {
              text: "Yes",
              onPress: async () => {
                try {
                  await api.post(
                    `/user-exercises/${userExerciseId}/confirm_decrease/`,
                    { confirm: "yes" }
                  );
                  Alert.alert(
                    "Success",
                    "Exercise difficulty decreased for next time!"
                  );
                  navigation.goBack();
                } catch (error) {
                  console.error("Error decreasing difficulty:", error);
                  Alert.alert(
                    "Error",
                    "Failed to decrease difficulty. Please try again."
                  );
                  navigation.goBack();
                }
              },
            },
          ]
        );
      } else {
        Alert.alert("Success", "Exercise completed successfully", [
          { text: "OK", onPress: () => navigation.goBack() },
        ]);
      }
    } catch (error) {
      // Handle error when saving exercise completion
      console.error("Error saving exercise:", error);
      Alert.alert(
        "Error",
        "Failed to save exercise completion. Please try again."
      );
    }
  };

  // Generate dropdown data
  const getDropdownData = (maxValue: number) => {
    return Array.from({ length: maxValue }, (_, i) => ({
      label: `${i + 1}`,
      value: i + 1,
    }));
  };

  // Generate pain level dropdown data
  const getPainLevelData = () => {
    return Array.from({ length: 11 }, (_, i) => ({
      label: `${i}`,
      value: i,
    }));
  };

  return {
    showCompletionForm,
    setShowCompletionForm,
    reps,
    setReps,
    sets,
    setSets,
    painLevel,
    setPainLevel,
    handleSave,
    getDropdownData,
    getPainLevelData,
  };
}
