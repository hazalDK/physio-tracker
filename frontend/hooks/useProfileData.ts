import { useState, useEffect } from "react";
import { Alert } from "react-native";
import * as SecureStore from "expo-secure-store";
import axios from "axios";
import { userProfileType } from "@/types/user";
import { useAuth } from "./useAuth";
import { useInjuryData } from "./useInjuryData";

/**
 * Custom hook to manage user profile data and related operations.
 * returns an object containing user profile data, loading state, modal visibility, and functions to handle profile updates.
 */
export function useProfileData() {
  const [userProfile, setUserProfile] = useState<userProfileType | null>(null);
  const [isModalVisible, setIsModalVisible] = useState(false);
  const [newUsername, setNewUsername] = useState("");
  const [newFirstName, setNewFirstName] = useState("");
  const [newLastName, setNewLastName] = useState("");
  const [newEmail, setNewEmail] = useState("");
  const [newDateOfBirth, setNewDateOfBirth] = useState<Date | undefined>(
    undefined
  );
  const { injuryTypes, loading: injuryLoading } = useInjuryData();
  const [open, setOpen] = useState(false);
  const [isUpdating, setIsUpdating] = useState(false);
  const [loading, setLoading] = useState(true);
  const { createApiInstance, refreshToken } = useAuth();

  // Function to fetch user profile data and handle token refresh if needed
  const fetchUserProfile = async () => {
    try {
      // Set loading state to true
      setLoading(true);
      // Create API instance
      const api = await createApiInstance();
      if (!api) return;
      try {
        // Fetch user profile data
        const response = await api.get(`/users/me/`);

        // set user profile data
        setUserProfile(response.data);
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) throw new Error("No refresh token available");

          // Create a new API instance with the refreshed token
          const api = await createApiInstance();
          if (!api) return;

          // Retry fetching user profile data
          const response = await axios.get(`${api}/users/me/`, {});

          if (response.status === 200 || response.status === 201) {
            // Set user profile data
            setUserProfile(response.data);
          } else {
            // Handle other errors
            console.error("API request failed:", error);
            Alert.alert(
              "Error",
              "Failed to load your profile. Please check your connection and try again."
            );
            setUserProfile(null);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          Alert.alert(
            "Error",
            "Failed to load your profile. Please check your connection and try again."
          );
          setUserProfile(null);
        }
      }
    } catch (error) {
      console.error("Failed to load exercise data:", error);
    } finally {
      setLoading(false);
    }
  };

  // Fetch user profile data on component mount
  useEffect(() => {
    fetchUserProfile();
  }, []);

  // Function to handle profile update
  // This function is called when the user submits the profile update form
  const handleEditProfile = async () => {
    // The button is disabled while the profile is being updated
    setIsUpdating(true);

    // Validate input fields to ensure at least one field is filled
    if (
      !newUsername &&
      !newFirstName &&
      !newLastName &&
      !newEmail &&
      !newDateOfBirth
    ) {
      Alert.alert("Error", "Please fill in at least one field");
      setIsUpdating(false);
      return;
    }

    // Basic email validation if email is provided
    if (newEmail && !validateEmail(newEmail)) {
      Alert.alert("Error", "Please enter a valid email address");
      setIsUpdating(false);
      return;
    }

    try {
      // Get the api instance
      const api = await createApiInstance();
      if (!api) return;

      // Build the updated data object with only the fields that are filled
      // This ensures that only the fields that the user wants to update are sent to the server
      const updatedData = {
        ...(newUsername && { username: newUsername }),
        ...(newFirstName && { first_name: newFirstName }),
        ...(newLastName && { last_name: newLastName }),
        ...(newEmail && { email: newEmail }),
        ...(newDateOfBirth && {
          date_of_birth: newDateOfBirth.toISOString().split("T")[0],
        }),
      };

      const response = await api.put(
        "http://192.168.68.111:8000/users/update_profile/",
        updatedData
      );

      if (response.status === 200 || response.status === 201) {
        // Handle success response from the API
        // This includes updating the user profile state and closing the modal
        Alert.alert("Success", "Profile updated successfully!");
        setIsModalVisible(false);
        setNewUsername("");
        setNewFirstName("");
        setNewLastName("");
        setNewEmail("");
        setNewDateOfBirth(undefined);
        fetchUserProfile();
      } else {
        Alert.alert("Error", "Unexpected response from the server");
      }
    } catch (error) {
      // Handle error response from the API
      // This includes logging the error and showing an alert to the user
      console.error("Error updating profile:", error);
      Alert.alert("Error", "Failed to update profile");
    } finally {
      setIsUpdating(false);
    }
  };

  // Function to validate email format
  const validateEmail = (email: string) => {
    const re = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return re.test(email);
  };

  // Function to toggle the visibility of the modal
  // This function is called when the user clicks the button to open or close the modal
  const toggleModal = () => {
    setIsModalVisible(!isModalVisible);

    // If opening the modal, initialize form fields with current profile values
    if (!isModalVisible && userProfile) {
      setNewUsername(userProfile.username || "");
      setNewFirstName(userProfile.first_name || "");
      setNewLastName(userProfile.last_name || "");
      setNewEmail(userProfile.email || "");
      // Date of birth is handled separately since it needs to be converted to a Date object
    }
  };

  return {
    userProfile,
    injuryTypes,
    loading,
    isModalVisible,
    newUsername,
    newFirstName,
    newLastName,
    newEmail,
    newDateOfBirth,
    open,
    isUpdating,
    setNewUsername,
    setNewFirstName,
    setNewLastName,
    setNewEmail,
    setNewDateOfBirth,
    setOpen,
    handleEditProfile,
    toggleModal,
  };
}
