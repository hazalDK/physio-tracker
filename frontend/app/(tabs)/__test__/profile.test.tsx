import React from "react";
import { render, fireEvent, act } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import { Alert } from "react-native";
import Profile from "../profile";
import { useProfileData } from "@/hooks/useProfileData";

// Mock the hooks
jest.mock("@/hooks/useProfileData");

// Mock DateTimePicker
jest.mock("@react-native-community/datetimepicker", () => {
  const mockComponent = ({
    onChange,
  }: {
    onChange: (event: { type: string }, date?: Date) => void;
  }) => {
    return (
      <button
        data-testid="date-picker-button"
        onClick={() => {
          return onChange({ type: "set" }, new Date("2000-01-01"));
        }}
      >
        Select Date
      </button>
    );
  };
  return mockComponent;
});

// Mock Alert
jest.spyOn(Alert, "alert").mockImplementation(() => {});

describe("Profile Component", () => {
  // Setup mock data for testing
  const mockUserProfile = {
    username: "testuser",
    first_name: "Test",
    last_name: "User",
    email: "test@example.com",
    date_of_birth: "1990-01-01",
    injury_type: 1,
  };

  const mockInjuryTypes = [
    { id: 1, name: "Sprain" },
    { id: 2, name: "Fracture" },
  ];

  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
  });

  // Update the first test to provide mock data
  it("renders correctly with data", () => {
    // Add mock implementation before rendering
    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: false,
      toggleModal: jest.fn(),
    });

    let tree;
    act(() => {
      tree = renderer.create(<Profile />).toJSON();
    });
    expect(tree).toMatchSnapshot();
  });

  it("renders loading state when loading is true", () => {
    (useProfileData as jest.Mock).mockReturnValue({
      loading: true,
    });

    const { getByTestId } = render(<Profile />);
    expect(getByTestId("loading-indicator")).toBeTruthy();
  });

  it("renders user profile data correctly", () => {
    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: false,
    });

    const { getByText } = render(<Profile />);

    expect(getByText("testuser")).toBeTruthy();
    expect(getByText("Test User")).toBeTruthy();
    expect(getByText("test@example.com")).toBeTruthy();
    expect(getByText("01/01/1990")).toBeTruthy();
    expect(getByText("Sprain")).toBeTruthy();
    expect(getByText("Edit Profile")).toBeTruthy();
  });

  it('renders "Not provided" for missing profile data', () => {
    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: {
        injury_type: 3,
      },
      injuryTypes: mockInjuryTypes,
      isModalVisible: false,
    });

    const { getAllByText } = render(<Profile />);
    const notProvidedTexts = getAllByText("Not provided");

    // Should have multiple "Not provided" texts for username, name, email, DOB
    expect(notProvidedTexts.length).toBeGreaterThan(0);

    // Should show "Unknown" for injury type not in the list
    expect(getAllByText("Unknown").length).toBe(1);
  });

  it("opens edit profile modal when button is pressed", () => {
    const toggleModal = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: false,
      toggleModal,
    });

    const { getByText } = render(<Profile />);
    fireEvent.press(getByText("Edit Profile"));

    expect(toggleModal).toHaveBeenCalledTimes(1);
  });

  it("renders modal with profile data when open", () => {
    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@example.com",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: false,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail: jest.fn(),
      setNewDateOfBirth: jest.fn(),
      setOpen: jest.fn(),
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
    });

    const { getByText, getByPlaceholderText, getByTestId } = render(
      <Profile />
    );

    expect(getByText("Update Profile")).toBeTruthy();
    expect(getByPlaceholderText("username").props.value).toBe("testuser");
    expect(getByPlaceholderText("first name").props.value).toBe("Test");
    expect(getByPlaceholderText("last name").props.value).toBe("User");
    expect(getByPlaceholderText("email").props.value).toBe("test@example.com");
    expect(getByTestId("datetime-text").props.children).toBe("01/01/1990");
  });

  it("handles text input changes in modal", () => {
    const setNewUsername = jest.fn();
    const setNewFirstName = jest.fn();
    const setNewLastName = jest.fn();
    const setNewEmail = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "",
      newFirstName: "",
      newLastName: "",
      newEmail: "",
      newDateOfBirth: null,
      open: false,
      isUpdating: false,
      setNewUsername,
      setNewFirstName,
      setNewLastName,
      setNewEmail,
      setNewDateOfBirth: jest.fn(),
      setOpen: jest.fn(),
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
    });

    const { getByPlaceholderText } = render(<Profile />);

    fireEvent.changeText(getByPlaceholderText("username"), "newusername");
    fireEvent.changeText(getByPlaceholderText("first name"), "New");
    fireEvent.changeText(getByPlaceholderText("last name"), "Name");
    fireEvent.changeText(getByPlaceholderText("email"), "new@example.com");

    expect(setNewUsername).toHaveBeenCalledWith("newusername");
    expect(setNewFirstName).toHaveBeenCalledWith("New");
    expect(setNewLastName).toHaveBeenCalledWith("Name");
    expect(setNewEmail).toHaveBeenCalledWith("new@example.com");
  });

  it("opens date picker when date field is pressed", () => {
    const setOpen = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@example.com",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: false,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail: jest.fn(),
      setNewDateOfBirth: jest.fn(),
      setOpen,
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
    });

    const { getByTestId } = render(<Profile />);

    fireEvent.press(getByTestId("date-picker-button"));
    expect(setOpen).toHaveBeenCalledWith(true);
  });

  it("clears date of birth when clear button is pressed", () => {
    const setNewDateOfBirth = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@example.com",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: false,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail: jest.fn(),
      setNewDateOfBirth,
      setOpen: jest.fn(),
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
    });

    const { getByText } = render(<Profile />);

    fireEvent.press(getByText("Clear"));
    expect(setNewDateOfBirth).toHaveBeenCalledWith(undefined);
  });

  it("calls handleEditProfile when update button is pressed", () => {
    const handleEditProfile = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@example.com",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: false,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail: jest.fn(),
      setNewDateOfBirth: jest.fn(),
      setOpen: jest.fn(),
      handleEditProfile,
      toggleModal: jest.fn(),
    });

    const { getByText } = render(<Profile />);

    fireEvent.press(getByText("Update"));
    expect(handleEditProfile).toHaveBeenCalledTimes(1);
  });

  it("disables inputs and buttons when updating", () => {
    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@example.com",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: true,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail: jest.fn(),
      setNewDateOfBirth: jest.fn(),
      setOpen: jest.fn(),
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
    });

    const { getByText, getByPlaceholderText } = render(<Profile />);

    expect(getByPlaceholderText("username").props.editable).toBe(false);
    expect(getByPlaceholderText("first name").props.editable).toBe(false);
    expect(getByPlaceholderText("last name").props.editable).toBe(false);
    expect(getByPlaceholderText("email").props.editable).toBe(false);
    expect(getByText("Updating...")).toBeTruthy();
  });

  it("handles email input changes in modal", () => {
    const setNewEmail = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@example.com",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: false,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail,
      setNewDateOfBirth: jest.fn(),
      setOpen: jest.fn(),
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
    });

    const { getByTestId } = render(<Profile />);

    fireEvent.changeText(getByTestId("email-input"), "updated@example.com");
    expect(setNewEmail).toHaveBeenCalledWith("updated@example.com");
  });

  it("shows error message for invalid email format", () => {
    const setNewEmail = jest.fn();
    const setEmailError = jest.fn();

    (useProfileData as jest.Mock).mockReturnValue({
      loading: false,
      userProfile: mockUserProfile,
      injuryTypes: mockInjuryTypes,
      isModalVisible: true,
      newUsername: "testuser",
      newFirstName: "Test",
      newLastName: "User",
      newEmail: "test@gmail",
      newDateOfBirth: new Date("1990-01-01"),
      open: false,
      isUpdating: false,
      setNewUsername: jest.fn(),
      setNewFirstName: jest.fn(),
      setNewLastName: jest.fn(),
      setNewEmail,
      setNewDateOfBirth: jest.fn(),
      setOpen: jest.fn(),
      handleEditProfile: jest.fn(),
      toggleModal: jest.fn(),
      setEmailError,
    });

    const { getByTestId, getByText } = render(<Profile />);

    fireEvent.changeText(getByTestId("email-input"), "invalid-email-format");
    expect(() =>
      getByText("Please enter a valid email (e.g., user@example.com).")
    ).toThrow();
  });
});
