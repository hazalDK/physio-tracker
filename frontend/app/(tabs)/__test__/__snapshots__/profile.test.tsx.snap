// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component renders correctly with data 1`] = `null`;
